framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,656 @@
+
+        // For each class
+        for (String clazz : classes) {
+            printClass(clazz, scene.getClasses().get(clazz), printWriter);
+        }
+        printWriter.flush();
+    }
+
+    /**
+     * Print the class body, or nothing if this is an anonymous inner class.
+     *
+     * @param classname the class name
+     * @param aClassWrapper the representation of the class
+     * @param printWriter the writer on which to print
+     */
+    private static void printClass(
+            @BinaryName String classname, AClassWrapper aClassWrapper, PrintWriter printWriter) {
+
+        String basename = basenamePart(classname);
+
+        if ("package-info".equals(basename) || "module-info".equals(basename)) {
+            return;
+        }
+
+        // Do not attempt to print stubs for anonymous inner classes or their inner classes, because
+        // the stub parser cannot read them. (An anonymous inner class has a basename like Outer$1,
+        // so this check ensures that no single class name is exclusively composed of digits.)
+        if (digitPattern.matcher(basename).matches()) {
+            return;
+        }
+
+        aClassWrapper.checkIfPrintable();
+
+        String innermostClassname =
+                basename.contains("$")
+                        ? basename.substring(basename.lastIndexOf('$') + 1)
+                        : basename;
+
+        String pkg = packagePart(classname);
+        if (!"".equals(pkg)) {
+            printWriter.println("package " + pkg + ";");
+        }
+
+        int curlyCount = printClassDefinitions(basename, aClassWrapper, printWriter);
+
+        String indentLevel = indents(curlyCount);
+
+        if (aClassWrapper.isEnum()) {
+            List<VariableElement> enumConstants = aClassWrapper.getEnumConstants();
+            if (enumConstants.size() != 0) {
+                StringJoiner sj = new StringJoiner(", ");
+                for (VariableElement enumConstant : enumConstants) {
+                    sj.add(enumConstant.getSimpleName());
+                }
+
+                printWriter.println(indentLevel + "// enum constants:");
+                printWriter.println();
+                printWriter.println(indentLevel + sj.toString() + ";");
+                printWriter.println();
+            }
+        }
+
+        printFields(aClassWrapper, printWriter, indentLevel);
+
+        if (aClassWrapper.getMethods().keySet().size() != 0) {
+            // print method signatures
+            printWriter.println(indentLevel + "// methods:");
+            printWriter.println();
+            for (Map.Entry<String, AMethodWrapper> methodEntry :
+                    aClassWrapper.getMethods().entrySet()) {
+                printMethodDeclaration(
+                        methodEntry.getValue(), innermostClassname, printWriter, indentLevel);
+            }
+        }
+        for (int i = curlyCount - 1; i >= 0; i--) {
```

Comment:
This code is correct, but it's nonstandard to count down, so I had to look twice to make sure it is correct and to try to understand the reason that you chose to count down here.  I suggest making it count up, which is a tiny bit shorter code too.

Reply:
I don't find counting up to be any clearer (nor any less clear...), but I've made the change, so you can judge.